From: Eric Dumazet <edumazet@google.com>
Subject: veth: extend device features
Git-commit: 8093315a91340bca52549044975d8c7f673b28a1
Patch-mainline: v3.9
References: bsc#879381
Acked-by: Jiri Bohac <jbohac@suse.cz>

[ For a SLE11-SP4 maintenance update, let's keep the the features off by default,
while allowing to turn them on via ethtool - jbohac ]

veth is lacking most modern facilities, like SG, checksums, TSO.

It makes sense to extend dev->features to get them, or GRO aggregation
is defeated by a forced segmentation.

Reported-by: Andrew Vagin <avagin@parallels.com>
Signed-off-by: Eric Dumazet <edumazet@google.com>
Cc: Michał Mirosław <mirq-linux@rere.qmqm.pl>
Signed-off-by: David S. Miller <davem@davemloft.net>

diff --git a/drivers/net/veth.c b/drivers/net/veth.c
index 6c0a3b0..e8562d5 100644
--- a/drivers/net/veth.c
+++ b/drivers/net/veth.c
@@ -258,6 +258,10 @@ static const struct net_device_ops veth_netdev_ops = {
 	.ndo_set_mac_address = eth_mac_addr,
 };
 
+#define VETH_FEATURES (NETIF_F_SG | NETIF_F_FRAGLIST | NETIF_F_ALL_TSO |    \
+		       NETIF_F_HW_CSUM | NETIF_F_RXCSUM | NETIF_F_HIGHDMA | \
+		       NETIF_F_HW_VLAN_TX | NETIF_F_HW_VLAN_RX)
+
 static void veth_setup(struct net_device *dev)
 {
 	ether_setup(dev);
@@ -267,9 +271,12 @@ static void veth_setup(struct net_device *dev)
 	dev->netdev_ops = &veth_netdev_ops;
 	dev->ethtool_ops = &veth_ethtool_ops;
 	dev->features |= NETIF_F_LLTX;
+/*	don't change the defaults in a maintenance update, but
+ *	allow the features to be set by ethtool */
+/*	dev->features |= VETH_FEATURES; */
 	dev->destructor = veth_dev_free;
 
-	dev->hw_features = NETIF_F_NO_CSUM | NETIF_F_SG | NETIF_F_RXCSUM;
+	dev->hw_features = VETH_FEATURES;
 }
 
 /*
